File: 2021/03/diagnose.py
def read_input():
    with open('../input/03.txt') as f:
        return f.read().split()

def consumption(input):
    transposed = [list(row) for row in zip(*input)]
    gamma = ''.join(['1' if row.count('1') > row.count('0') else '0' for row in transposed])
    epsilon = ''.join(['1' if row.count('1') < row.count('0') else '0' for row in transposed])
    return int(gamma, 2) * int(epsilon, 2)

def oxygen(input, index = 0):
    if len(input) == 1:
        return input[0]
    else:
        ones = list(filter(lambda n: n[index] == '1', input))
        zeros = list(filter(lambda n: n[index] == '0', input))
        return oxygen(ones if len(ones) >= len(zeros) else zeros, index + 1)

def co2(input, index = 0):
    if (len(input) == 1):
        return input[0]
    else:
        ones = list(filter(lambda n: n[index] == '1', input))
        zeros = list(filter(lambda n: n[index] == '0', input))
        return co2(zeros if len(zeros) <= len(ones) else ones, index + 1)

def life_support(input):
    return int(oxygen(input), 2) * int(co2(input), 2)

def main():
    input = read_input()
    print(consumption(input))
    print(life_support(input))

if __name__ == "__main__":
    main()

File: 2021/03/diagnose.ts
import * as fs from 'fs';

const input = fs.readFileSync('../input/03.txt').toString().split('\n')

const count = (line: string[], value: '1' | '0'): number => {
    return line.filter(c => c == value).length
}

const gamma = (matrix: string[][]): number => {
    const binary = matrix.map((line) => count(line, '1') > count(line, '0') ? '1' : '0').join('')

    return parseInt(binary, 2)
}

const epsilon = (matrix: string[][]): number => {
    const binary = matrix.map((line) => count(line, '1') < count(line, '0') ? '1' : '0').join('')

    return parseInt(binary, 2)
}

const consumption = (input: string[]): number => {
    const transposed = input[0].split('').map((col, i) => input.map((row) => row[i]))
    
    return gamma(transposed) * epsilon(transposed)
}

const oxygen = (input: string[], index: number = 0): string => {
    if (input.length == 1) {
        return input[0]
    } else {
        const ones = input.filter(v => v[index] == '1')
        const zeros = input.filter(v => v[index] == '0')

        return ones.length >= zeros.length ? oxygen(ones, index + 1) : oxygen(zeros, index + 1)
    }
}

const co2 = (input: string[], index: number = 0): string => {    
    if (input.length == 1) {
        return input[0]
    } else {
        const ones = input.filter(v => v[index] == '1')
        const zeros = input.filter(v => v[index] == '0')        

        return ones.length < zeros.length ? co2(ones, index + 1) : co2(zeros, index + 1)
    }
}

const lifeSupport = (input: string[]) : number => {
    return parseInt(oxygen(input), 2) * parseInt(co2(input), 2)
}

console.log(consumption(input))
console.log(lifeSupport(input))

